src/Features/Core/Portable/Completion/Providers/AbstractTypeImportCompletionProvider.cs

```diff
@@ -0,0 +1,175 @@
+            SemanticModel semanticModel,
+            ImmutableHashSet<string>.Builder builder,
+            CancellationToken cancellationToken);
+
+        public override async Task ProvideCompletionsAsync(CompletionContext completionContext)
+        {
+            var document = completionContext.Document;
+            var cancellationToken = completionContext.CancellationToken;
+            var workspace = document.Project.Solution.Workspace;
+            var experimentationService = workspace.Services.GetService<IExperimentationService>();
+
+            var importCompletionOptionValue = completionContext.Options.GetOption(CompletionOptions.ShowImportCompletionItems, document.Project.Language);
+
+            // Don't trigger import completion if the option value is "default" and the experiment is disabled for the user. 
+            if (importCompletionOptionValue == false ||
+                (importCompletionOptionValue == null && experimentationService?.IsExperimentEnabled(WellKnownExperimentNames.TypeImportCompletion) != true))
+            {
+                return;
+            }
+
+            var syntaxContext = await CreateContextAsync(document, completionContext.Position, cancellationToken).ConfigureAwait(false);
+            if (!syntaxContext.IsTypeContext)
+            {
+                return;
+            }
+
+            using (Logger.LogBlock(FunctionId.Completion_TypeImportCompletionProvider_GetCompletionItemsAsync, cancellationToken))
+            {
+                var project = document.Project;
+                var typeImportCompletionService = workspace.Services.GetService<ITypeImportCompletionService>();
+
+                // Find all namespaces in scope at current cursor location, 
+                // which will be used to filter so the provider only returns out-of-scope types.
+                var namespacesInScope = GetNamespacesInScope(document, syntaxContext, cancellationToken);
+                Action<CompletionItem> handleAccessibleItem = item => AddItems(item, completionContext, namespacesInScope);
+
+                // Get completion items from current project.
+                await typeImportCompletionService.GetAccessibleTopLevelTypesFromProjectAsync(project, handleAccessibleItem, cancellationToken)
+                    .ConfigureAwait(false);
+
+                var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+                foreach (var reference in compilation.References)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    var declarationsInReference = ImmutableArray<CompletionItem>.Empty;
+                    if (reference is CompilationReference compilationReference)
+                    {
+                        // Get completion items from source references.
+                        await typeImportCompletionService.GetAccessibleTopLevelTypesFromCompilationReferenceAsync(
```

Comment:
also, I think GetAccessibleTopLevelTypesFromCompilationReferenceAsync and GetAccessibleTopLevelTypesFromProjectAsync could be combined to 1.

like GetAccessibleTopLevelTypesFromAsync(Project from, Project to, ....)

Reply:
Done